lk2nd/smp/cpu-boot.c

```diff
@@ -56,6 +61,57 @@ int qcom_set_boot_addr(uint32_t addr, bool arm64)
 				QCOM_SCM_BOOT_FLAG_COLD_ALL, addr);
 }
 
+void qcom_power_on_l2_cache(uint32_t mpidr, uint32_t base)
+{
+	uint32_t val;
+
+	/* Skip if cluster L2 is already powered on */
+	val = readl(base + L2_PWR_STATUS);
+	if (val & BIT(9)) {
+		dprintf(INFO, "CPU (%d) L2 cache already powered-up\n", mpidr);
+		return;
+	}
+
+	/* Close L2/SCU Logic GDHS and power up the cache */
+	writel(0x10D700, base + L2_PWR_CTL);
+
+	/* Assert PRESETDBGn */
+	writel(0x400000, base + L2_PWR_CTL_OVERRIDE);
+	dsb();
+	udelay(2);
+
+	/* De-assert L2/SCU memory Clamp */
+	writel(0x101700, base + L2_PWR_CTL);
+
+	/* Wakeup L2/SCU RAMs by deasserting sleep signals */
+	writel(0x101703, base + L2_PWR_CTL);
+	dsb();
+	udelay(2);
+
+	/* Enable clocks via SW_CLK_EN */
+	writel(0x01, base + L2_CORE_CBCR);
+
+	/* De-assert L2/SCU logic clamp */
+	writel(0x101603, base + L2_PWR_CTL);
+	dsb();
+	udelay(2);
+
+	/* De-assert PRESSETDBg */
+	writel(0x0, base + L2_PWR_CTL_OVERRIDE);
+
+	/* De-assert L2/SCU Logic reset */
+	writel(0x100203, base + L2_PWR_CTL);
+	dsb();
+	udelay(54);
+
+	/* Turn on the PMIC_APC */
+	writel(0x10100203, base + L2_PWR_CTL);
```

Comment:
Can you rewrite this with proper register definitions, similar to `qcom_power_up_arm_cortex()` below? They are publicly documented in the APQ8016E TRM and/or the HRD. This will also allow you to drop most of the comments actually, because `|= L2_PWR_CTL_PMIC_APC_ON` should be as clear as `Turn on the PMIC_APC`.

Reply:
Sure NP, I think I can find a copy

Reply:
There is no need to search for a copy, you can easily download it from the Qualcomm web page. :)

- TRM: https://developer.qualcomm.com/download/sd410/snapdragon-410e-technical-reference-manual.pdf, page 97
- HRD: https://developer.qualcomm.com/download/sd410/hardware-register-description-qualcomm-snapdragon-410.pdf